### PR TITLE
Add multiple recipients on the expenditure creation page

### LINKED ISSUE #3402: Expenditures: As a User I can add multiply recipients
[Provided by Michał via Contractors project management platform]

### Description
User should be able to add multiply recipients with all required data

#### Acceptance Criteria

- an option to select the recipient (autocomplete)
- an option to set up values
- an option to set up claim delay (with tooltip)

### Design

https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2420%3A1079289

2453b3a6bb3092bbfd15735b22b75752

Screenshot 2022-05-17 152041

2455bf670091c7210701e6217ea667cd

2459c7f3a01a2bd283ede17c12974086

### PR DESCRIPTION
**Pull request by agnieszkajarosikloj**
## Description
Expenditure form section were you can add multiple recipients with all required data.

[Figma design](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=2420%3A1079289)

**New stuff** ✨

* UserPickerWithSearch component - it has functionality almost the same as SingleUserPicker 
but its appearance is adapted to the advanced expenditure design
Screenshot 2022-05-26 at 15 43 16

* Payment component - you can add here multiple recipients
* Recipient component - form section with all required data
* 
Screenshot 2022-05-26 at 15 44 01

TODO: 
- add actual data to components (SingleUserPicker component - data and Input component - maxAmount)

Resolves #3402

### CODE REVIEW

File: src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    <div className={styles.paymentContainer}>
+      <div className={styles.recipientContainer}>
+        <div className={styles.payments}>
+          <FormattedMessage {...MSG.payments} />
+        </div>
+        <FieldArray
+          name="recipients"
+          render={({ push, remove }) => (
+            <>
+              {recipients.map((recipient, index) => (
```
Comment (arrenv) [inline src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx:55]: We should be able to get the members just like we do with the `CreatePaymentDialog` component. Check the usage of `useMembersSubscription`. It may make this easier then using mock data here.

**Review by arrenv — CHANGES_REQUESTED** [review]
@agnieszkajarosikloj A good addition, thank you. I have added a comment regarding getting Members as well as some comments below.

1. Match the amount input padding and font size.

d9dee9e1-5b24-4e0c-8e69-e51387ed55f1

2. Right align the font and ensure the size matches the design in the Claim delay input.
 
63194fa4-172a-4816-8863-a74f127dfc15

3. Make the `Another token` button clickable to add another token input and select option within the same recipient section, fix the padding as of the design. 

download (7)

4. Ensure the width of the period select stays the same when changing the selection.

download (8)

5. Show the hover state when also hovering over the drop down arrow. Try to also improve how the icon renders.

5142e19d-3160-4bf4-9348-f5c4e182395d

6. Please add a title tag to the delete icon with "Delete recipient".

download (10)

7. Please add a separate scroll to the left aside.

bf01af03-8cab-4c92-b3e2-60947cdd7bd9

**Review by chinins — COMMENTED** [review]
In addition to Arren's comments.

@arrenv, should the profile icon that opens the recipient field also close it?

File: src/img/icons/plus.svg [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<svg width="9" height="9" viewBox="0 0 9 9"  xmlns="http://www.w3.org/2000/svg">
```
[the remark this answers is not in the document]
Comment (arrenv) [reply src/img/icons/plus.svg:1]: @chinins good pickup, I think we should keep the consistency here. @agnieszkajarosikloj please make the `viewbox` `0 0 30 30` let me know if you want me to export this for you.

**Comment by arrenv** [comment]
> @arrenv, should the profile icon that opens the recipient field also close it?

I thought it felt weird. Yes, it would be better to have the icon both open and close (toggle) the recipient field.

**Review by arrenv — CHANGES_REQUESTED** [review]
Thank you @agnieszkajarosikloj.

1. it looks like there is still an issue with the alignment.

2c02a001-c228-4608-94fd-05d3ee0ab18d

Design

Screenshot 2022-06-08 151938

2. The select recipient is acting a little weird, it seems each element behaves differently when clicked on, when there is a user selected. The whole element showed work like how it does when the down caret is clicked, i.e. opens the select without removing the user.

select-recipient-issue

3. The icon still looks cut off.

a5e851d5-c861-4118-9f67-f2bca5fa49c8

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                        name={`recipients[${index}].value[${idx}].amount`}
+                        appearance={{
+                          theme: 'underlined',
+                          size: 'small',
+                        }}
+                        label={MSG.defaultValueLabel}
+                        placeholder="Not set"
+                        formattingOptions={{
+                          numeral: true,
+                          // @ts-ignore
```
Comment (rdig) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx:140]: Why is this here? Please leave a comment explaining it

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+                            <FormattedMessage {...MSG.removeTokenText} />
+                          </Button>
+                        )}
+                      </div>
+                      <div>
+                        <TokenSymbolSelector
+                          label=""
+                          tokens={tokensData}
+                          // eslint-disable-next-line max-len
+                          name={`recipients[${index}].value[${idx}].tokenAddress`}
```
Comment (rdig) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx:170]: Nesting and accessors here are getting out of hand really...

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                          <FormattedMessage {...MSG.valueError} />
+                        </div>
+                      )}
+                      {/* if last */}
+                      {tokens.length === idx + 1 && (
+                        <Button
+                          type="button"
+                          onClick={() => arrayHelpers.push(newToken)}
+                          appearance={{ theme: 'blue' }}
+                          style={{ margin: '7px 0' }}
```
Comment (rdig) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx:186]: Please use css classes to change styles

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+                  trigger="hover"
+                  placement="right-start"
+                >
+                  <Icon name="question-mark" className={styles.questionIcon} />
+                </Tooltip>
+              </div>
+
+              <div className={styles.delayControlsContainer}>
+                <Input
+                  name={`recipients[${index}].delay.amount`}
```
Comment (rdig) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx:223]: Not really a big fan of the format for naming these form elements

File: src/modules/pages/components/ExpenditurePage/ExpenditurePage.css [hunk 7]
```diff
@@ -20,5 +20,17 @@
 .sidebar {
   height: calc(100vh - navBarHeight);
   width: sidebarWidth;
+  overflow-y: auto;
   background-color: var(--grey-blue-3);
 }
+
+.sidebar::-webkit-scrollbar {
+  display: none;
+}
```
Comment (rdig) [inline src/modules/pages/components/ExpenditurePage/ExpenditurePage.css:29]: We don't need to add vendor suffixes manually, as webpack runs a `postcss` step for use which does this automatically for the bundled code

**Review by rdig — CHANGES_REQUESTED** [review]
Mostly good, however the core components are repeating too much code and need to be refactored.

Preferably just use the current core components and extend them to include the functionality you need

File: src/modules/core/components/OmniPicker/AdditionalOmniPicker.tsx [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+  select: (idx: number) => void;
+
+  /** @ignore Will be injected by `withOmniPicker` */
+  selected: number;
+
+  /** @ignore Will be injected by `withOmniPicker` */
+  triggerElement: HTMLElement | null;
+}
+
+class AdditionalOmniPicker extends Component<Props> {
```
[the remark this answers is not in the document]
Comment (agnieszkajarosikloj) [reply src/modules/core/components/OmniPicker/AdditionalOmniPicker.tsx:105]: I didn't modify the existing components as you suggested in an earlier comment in this PR: https://github.com/JoinColony/colonyDapp/pull/3393
Screenshot 2022-06-10 at 10 16 00

File: src/modules/core/components/UserPickerWithSearch/UserPickerWithSearch.tsx [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+  valueDataTest?: string;
+
+  sidebarRef?: HTMLElement | null;
+}
+
+interface EnhancedProps extends Props, WrappedComponentAdditionalProps {}
+
+const displayName = 'UserPickerWithSearch';
+
+const UserPickerWithSearch = ({
```
[the remark this answers is not in the document]
Comment (agnieszkajarosikloj) [reply src/modules/core/components/UserPickerWithSearch/UserPickerWithSearch.tsx:102]: I agree it will reduce code repetition, but like in the previous comment, you suggested me not to change existing core components: https://github.com/JoinColony/colonyDapp/pull/3393#discussion_r879709290

File: src/modules/pages/components/ExpenditurePage/ExpenditurePage.tsx [hunk 10]
```diff
@@ -1,21 +1,71 @@
...
+const initialValues = {
+  expenditure: 'advanced',
+  recipients: [newRecipient],
+};
+
+const validationSchema = yup.object().shape({
+  expenditure: yup.string().required(),
+  recipients: yup.array(
+    yup.object().shape({
+      recipient: yup.object().required('User is required'),
```
[the remark this answers is not in the document]
Comment (agnieszkajarosikloj) [reply src/modules/pages/components/ExpenditurePage/ExpenditurePage.tsx:24]: `required(() => MSG.userRequiredError)`
doesn't work here, I have an error: 
Screenshot 2022-06-10 at 12 32 30
I used a Form component here, not an ActionForm.

File: src/modules/core/components/OmniPicker/AdditionalOmniPicker.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (rdig) [reply src/modules/core/components/OmniPicker/AdditionalOmniPicker.tsx:105]: Ufff. Fine.

Leave like it is then. Disregard my comments related to extending existing core components.

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx (hunk 4 above)
Comment (agnieszkajarosikloj) [reply src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx:170]: There are two nested arrays in this form, the first with recipients, the second with values. That's why names look like that.
Screenshot 2022-06-10 at 12 37 18

File: src/modules/pages/components/ExpenditurePage/ExpenditurePage.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (rdig) [reply src/modules/pages/components/ExpenditurePage/ExpenditurePage.tsx:24]: This is not to do with the Form/ActionForm components, but with the way it's rendering the error message, which in this case is `FieldArray`, which I think causes the above issue.

There's two ways to go about this: extract the message descriptor string _(via something like `formatMessage`)_, or wrap it inside a react element.

```tsx
import { useIntl } from 'react-intl';

// inside the component
const { formatMessage, formatNumber } = useIntl();

// then you can pass it along as an argument to validation schema and use that to parse the message descriptor for the error message
validationSchema={validationSchema(formatMessage)}

// then, outside the component's scope, and inside the validator function
required(() => formatMessage(MSG.userRequiredError))
```

Or, just wrap it inside a react component
```tsx
required(() => <FormattedMessage {...MSG.userRequiredError} />)
```

**Review by arrenv — APPROVED** [review]
Thank @agnieszkajarosikloj for all the updates, looks good to me.

File: src/modules/core/components/OmniPicker/AdditionalOmniPicker.tsx [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+      onChange(evt);
+    }
+  };
+
+  handleFocus = (evt: SyntheticEvent<HTMLInputElement>) => {
+    const { onFocus } = this.props;
+    if (typeof onFocus == 'function') {
+      onFocus(evt);
+    }
+  };
```
Comment (rdig) [inline src/modules/core/components/OmniPicker/AdditionalOmniPicker.tsx:182]: Just a tough though, I know these were copy - pasted from the parent component, but I think theses can actually be refactored as a hook.

Something like:
```tsx
// The hook
enum Handle {
  OPEN = 'open',
  CLOSE = 'close',
  RESET = 'reset',
  ...
}

interface Handlers {
  [key: string]?: (evt: SyntheticEvent<HTMLInputElement>) => void;
}

const useHandleFactory = <D>(handles: Handle[], dependencies: D): Handlers => {
  const allHandlers: Handlers = {};
  const [props] = dependencies;

  handles.map((handle: Handle) => {
    const callbackName = `on${handle.charAt(0).toUpperCase()}${handle.slice(1)}`;
    const handlerName = `handle${callbackName.charAt(0).toUpperCase()}${callbackName.slice(1)}`;
    allHandlers[handlerName] = (evt: SyntheticEvent<HTMLInputElement>) => {
      const handlerCallback = props[callbackName];
      if (typeof handlerCallback == 'function') {
        handlerCallback(evt);
      }
    };
  });
};

// Then you would call it something like this (inside the component)

const { handleOnOpen, handleOnClose, handleOnReset, ... } = useHandleFactory(
  [Handle.OPEN, Handle.CLOSE, Handle.RESET, ... ],
  [this.props],
);
```

Of course, this is all theoretic only, I don't expect you to spend the time to actually implement this.

File: src/modules/core/components/UserPickerWithSearch/UserPickerWithSearch.tsx [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+import Button from '~core/Button';
+
+type AvatarRenderFn = (
+  address: Address,
+  user?: ItemDataType<AnyUser>,
+) => ReactNode;
+
+const MSG = defineMessages({
+  selectMember: {
+    id: 'SingleUserPicker.selectMember',
```
Comment (rdig) [inline src/modules/core/components/UserPickerWithSearch/UserPickerWithSearch.tsx:30]:
```suggestion
    id: 'UserPickerWIthSearch.selectMember',
```

This applies to all message descriptor id's

File: src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+import { newRecipient } from './constants';
+import {
+  useColonyFromNameQuery,
+  useMembersSubscription,
+} from '~data/generated';
+import { SpinnerLoader } from '~core/Preloaders';
+
+const MSG = defineMessages({
+  payments: {
+    id: 'dashboard.Expenditures.Payments.defaultPayment',
```
Comment (rdig) [inline src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx:23]:
```suggestion
    id: 'dashboard.ExpenditurePage.Payments.defaultPayment',
```

This applies to all other message descriptor id's on this page

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+import { AnyUser } from '~data/index';
+import { Address } from '~types/index';
+import { Recipient as RecipientType } from '../Payments/types';
+import { tokensData } from './constants';
+
+import styles from './Recipient.css';
+
+const MSG = defineMessages({
+  defaultRecipientLabel: {
+    id: 'dashboard.Expenditures.Recipient.defaultRecipientLabel',
```
Comment (rdig) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx:30]:
```suggestion
    id: 'dashboard.ExpenditurePage.Recipient.defaultRecipientLabel',
```

Again ,this applies to all other message descriptor id's on this page

**Review by rdig — APPROVED** [review]
This is all good. Very nice work @agnieszkajarosikloj :100: 

The only thing I could find were some mis-labeled message descriptors, but other than that, it's good to go as far as I'm concerned.

File: src/modules/core/components/OmniPicker/AdditionalOmniPicker.tsx [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+  /** @ignore Will be injected by `withOmniPicker` */
+  select: (idx: number) => void;
+
+  /** @ignore Will be injected by `withOmniPicker` */
+  selected: number;
+
+  /** @ignore Will be injected by `withOmniPicker` */
+  triggerElement: HTMLElement | null;
+}
+
```
Comment (chinins) [inline src/modules/core/components/OmniPicker/AdditionalOmniPicker.tsx:104]: This component is missing displayName, please, add it. It should be just `OmniPicker.AdditionalOmniPicker`.
I know that the initial component doesn't have it but it's an oversight on our part.

File: src/modules/core/components/UserPickerWithSearch/Dropdown.tsx [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+import * as ReactDOM from 'react-dom';
+
+import styles from './Dropdown.css';
+
+interface Props {
+  element: HTMLDivElement | null;
+  scrollContainer?: Window | HTMLElement | null;
+  children: React.ReactNode;
+}
+
```
Comment (chinins) [inline src/modules/core/components/UserPickerWithSearch/Dropdown.tsx:11]: Same here - missing the displayName. Should be `UserPickerWithSearch.Dropdown`

File: src/modules/dashboard/components/ExpenditurePage/Payments/Payments.css [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+  position: relative;
+}
+
+.signWrapper {
+  height: 30px;
+  width: 30px;
+  position: absolute;
+
+  /* negative margin is used to set the position of the button,
+  which collapse and expand singe recipient fields,
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Payments/Payments.css:41]: It's a nitpick but there are types here. I assume it should be `which collapses and expands single recipient fields`?

File: src/modules/dashboard/components/ExpenditurePage/Payments/Payments.css [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+  cursor: pointer;
+}
+
+.verticalDivider {
+  height: calc(100% - 124px);
+  position: absolute;
+  top: 56px;
+
+  /* negative margin is used to set the position of the vertical line
+  on the left side of the form, similar to margin in singWrapper class */
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Payments/Payments.css:56]: Should it be `signWrapper`?

File: src/modules/dashboard/components/ExpenditurePage/Payments/Payments.css [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+  align-items: center;
+}
+
+.circlePlusIcon {
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  margin-right: 2px;
+
+  /* negative margin is used here to adjust placement all row */
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Payments/Payments.css:84]: Not sure I understand the comment: `to adjust the placement of` what?

File: src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx (hunk 13 above)
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx:23]: The last part of id `defaultPayment` is not the same os the key `payment`

File: src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+    defaultMessage: 'Add payments',
+  },
+  recipient: {
+    id: 'dashboard.Expenditures.Payments.defaultRrecipient',
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx:27]: Same here

File: src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+                            title={MSG.plusIconTitle}
+                          />
+                        )}
+                        {index + 1}: <FormattedMessage {...MSG.recipient} />
+                        {recipients.length > 1 && (
+                          <Icon
+                            name="trash"
+                            className={styles.deleteIcon}
+                            onClick={() => remove(index)}
+                            title="Delete recipient"
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Payments/Payments.tsx:122]: It's better to use message descriptor as title, like you have for the expand icon. Please, update.

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.css [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  padding: 6px 0;
+}
+
+.delay {
+  display: flex;
+  align-items: center;
+  font-size: 13px;
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.css:52]: Please, use variable for font-sizes. In this case `--size-smallish`

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.css [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+
+.delayControlsContainer button {
+  min-width: 96px;
+}
+
+.tokenWrapper {
+  display: flex;
+  flex-direction: column;
+  align-items: flex-end;
+  font-size: 12px;
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.css:85]: Please, use variable for font-sizes. In this case `--size-small`

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.css [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+  padding: 6px 0;
+}
+
+.delay {
+  display: flex;
+  align-items: center;
+  font-size: 13px;
+  font-weight: var(--weight-bold);
+  color: var(--text);
+  letter-spacing: 0.5px;
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.css:55]: We have a variable `--spacing-medium`

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+  addTokenText: {
+    id: 'dashboard.Expenditures.Recipient.addTokenText',
+    defaultMessage: 'Another token',
+  },
+  removeTokenText: {
+    id: 'dashboard.Expenditures.Recipient.removeTokenText',
+    defaultMessage: 'Discard',
+  },
+  hoursLabel: {
+    id: 'dashboard.Expenditures.Recipient.daysOptionLabel',
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx:58]: Same id as the next item, should be `daysOptionLabel` at the end

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx [hunk 26]
```diff
@@ -0,0 +1,276 @@
...
+    defaultMessage: 'hours',
+  },
+  daysLabel: {
+    id: 'dashboard.Expenditures.Recipient.daysOptionLabel',
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx:62]: Id & object key don't match, please, update

File: src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx [hunk 27]
```diff
@@ -0,0 +1,276 @@
...
+    defaultMessage: 'days',
+  },
+  monthsLabel: {
+    id: 'dashboard.Expenditures.Recipient.monthsOptionLabel',
```
Comment (chinins) [inline src/modules/dashboard/components/ExpenditurePage/Recipient/Recipient.tsx:66]: Id & object key don't match, please, update

**Review by chinins — APPROVED** [review]
That's a huge PR - very nice work. There are several small things to correct as per my comments. I am pre-approving as they are very small.

**Review by chinins — APPROVED** [review]
Very nice work! I left some comments but pre-approving as they are very small.
